docs/docs/architecture/client_broker_protocol.md

```diff
@@ -0,0 +1,1223 @@
+  Each `Subscription` type contains these fields:
+
+  - `sId`: This integer represents a unique identifier for this subscription.
+    This integer needs to be unique only between this client and the broker.  A
+    simple incrementing
+    [variable](https://github.com/bloomberg/blazingmq/blob/9b692fe25f74543e954a27e30b6b15b0ae057c8d/src/groups/bmq/bmqt/bmqt_subscription.cpp#L44-L49)
+    can be used for generating unique `sId`s.
+
+  - `expression`: This field, which is of type
+    [`Expression`](https://github.com/bloomberg/blazingmq/blob/9b692fe25f74543e954a27e30b6b15b0ae057c8d/src/groups/bmq/bmqp/bmqp_ctrlmsg.xsd#L412C4-L426)
+    contains the actual subscription expression as a string type as well as
+    expression grammar version.
+
+  - `consumers`: This field is an array of type
+    [`ConsumerInfo`](https://github.com/bloomberg/blazingmq/blob/9b692fe25f74543e954a27e30b6b15b0ae057c8d/src/groups/bmq/bmqp/bmqp_ctrlmsg.xsd#L428-L453),
+    and represents various attributes related a subscription.
+
+> [!NOTE]
+> As previously mentioned, opening a queue in BlazingMQ is a two-step operation
+> -- the client needs to send `OpenQueue` request, followed by
+> `ConfigureStream` request.  At the API level, both of these steps should be
+> hidden by simply exposing someting like `openQueue` API, similar to the
+> [C++](https://github.com/bloomberg/blazingmq/blob/9b692fe25f74543e954a27e30b6b15b0ae057c8d/src/groups/bmq/bmqa/bmqa_session.h#L819-L840)
+> and
+> [Java](https://github.com/bloomberg/blazingmq-sdk-java/blob/ed0eb848f5ac2897cfe32cc481575d689a4cb1c2/bmq-sdk/src/main/java/com/bloomberg/bmq/Queue.java#L24-L34)
+> SDKs.
+
+### Posting Messages (Producers)
+
+Let's assume that a producer application opened a BlazingMQ queue by following
+the steps mentioned in the [above](#opening-a-queue) section.  The next step
+for the producer application would be to post some messages (PUT messages) on
+the queue.
+
+#### Post API
+
+Both C++ and Java SDKs support batching of messages i.e., producer application
+can send a batch of messages in one shot instead of sending them one by one.
+Batching can have a significant impact on the performance.  Additionally, in
+both SDKs, the `post()` API is asynchronous i.e., the API does not wait for the
+BlazingMQ broker to respond with ACK messages. In fact, in some cases, the API
+does not write the PUT messages to the TCP socket.  Instead, the SDK just
+accepts the messages, adds them to an internal buffer, and sends them later.
+This can occur if TCP layer is enforcing some flow control, or in case SDK is
+not connected to the broker, etc.
+
+Keeping above complications aside, the simplest way to send PUT messages is to
+do what [Java
+SDK](https://github.com/bloomberg/blazingmq-sdk-java/blob/ed0eb848f5ac2897cfe32cc481575d689a4cb1c2/bmq-sdk/src/main/java/com/bloomberg/bmq/Queue.java#L109-L115)
+is doing.  A public class called `PutMessage` could be created, with data
```

Comment:
```suggestion
does.  A public class called `PutMessage` could be created, with data
```

Reply:
Also, the shift in perspective from describing what the C++/Java SDKs do factually to describing what a new client SDK could do with C-ish syntax was jarring, especially in the middle of the "BlazingMQ Client Server Interaction" section.